cannon/mipsevm/page.go

```diff
@@ -1,12 +1,39 @@
 package mipsevm
 
 import (
+	"bytes"
+	"compress/zlib"
+	"encoding/base64"
 	"encoding/hex"
+	"encoding/json"
 	"fmt"
+	"io"
+	"regexp"
+	"sync"
 
 	"github.com/ethereum/go-ethereum/crypto"
 )
 
+var zlibWriterPool = sync.Pool{
+	New: func() any {
+		var buf bytes.Buffer
+		return zlib.NewWriter(&buf)
+	},
+}
+
+func detectEncoding(data []byte) (string, error) {
+	// Regular expressions to check for base64 and hex patterns
+	base64Pattern := "^[A-Za-z0-9+/]*={0,2}$"
+	hexPattern := "^[0-9a-fA-F]*$"
+
+	if regexp.MustCompile(hexPattern).MatchString(string(data)) {
+		return "hex", nil
+	} else if regexp.MustCompile(base64Pattern).MatchString(string(data)) {
+		return "base64", nil
+	}
```

Comment:
There is an issue where were hex chars are a strict subset of base64 so while its unlikely for memory, it's possible that we wrote base64 using only valid hex chars and would then interpret it wrong on reload.  That would cause games to be lost incorrectly so isn't something we can really risk. (Assuming I'm reading the regex right and the trailing `=` chars on base64 are optional).

We might be better off prefixing base64 content with `base64:` so there's no doubt about the formatting of the data.

Reply:
yea, it's a concern. It's also a problem in the other direction though - if a page is all 0s, it will be reported as base64 (which is why the hex check is first).

TBH I think we should migrate to this new format rather than supporting both, but for the short term supporting both is nice in terms of `prestate.json`

Reply:
Yeah agreed we don't want to live with hex for long, but having a clear `base64:` prefix would simplify the check and completely avoid any risk of incorrect detection while preserving compatibility for the existing prestate. I just think it's worth avoiding ever adding code that would potentially lead to us making invalid claims.